Add support for dynamically located Points of Interest

rafaellehmkuhl opened the pull request:
# Live-tracked Points of Interest


https://github.com/user-attachments/assets/f49a4a2c-3b0f-4261-911a-d65f23567500



## What this introduces

Points of Interest (POIs) can now follow a moving target instead of being pinned to a fixed spot.

- **Static or live coordinates** — a POI's latitude/longitude can be a fixed number (as before) or a data-lake expression such as `{{ mavlink/buoy/latitude }}`. Live POIs move on the map as their underlying values update.
- **POI manager** — a static/live toggle, plus a combobox to pick any number-typed data-lake variable (or type a custom expression). Live POIs that have no valid value yet are shown dimmed with a "Coordinates unknown" tooltip.
- **Tools → Map menu** — lists every POI with its live/static status and lets you center, edit or delete each one.
- **Center-on dial** — the map's "center on" speed dial now includes POIs: single-click to center, double-click to keep tracking. Home/vehicle/mission tracking is unchanged.
- **Map widget, mission map, HUD and edge arrows** all render the resolved POI positions and update live.

## Key design choices

- POI logic lives in a `usePointsOfInterest` composable plus small plain-`.ts` modules, instead of the mission Pinia store.
- A clean split between the **persisted definition** (name, icon, color, coordinate source, fallback) and the **resolved runtime POI** (current coordinates + live status) that the UI consumes. No runtime/derived state is ever written back into the persisted object.
- Display concerns (marker color, opacity, tooltip, status text) are centralized in `libs/utils-poi.ts` and shared by every consumer, removing the duplicated marker/tooltip code that previously lived in both map components.

## How POIs relate to the data lake

Every POI exposes its position through the data lake as two variables:

```
cockpit/pois/<id>/latitude
cockpit/pois/<id>/longitude
```

These are the single source of truth for a POI's *live* position, which keeps POIs consistent with Cockpit's broader move toward sourcing widget data from the data lake.

- **Static coordinate** — the fixed number is written into the backing variable once.
- **Live coordinate** — the expression is evaluated (via `evaluateDataLakeExpression`, so arithmetic like `{{ x }} * 10` works), and the result is re-written whenever any referenced variable changes. Each backing variable subscribes only to the variables its expression actually references, so updates are pushed, not polled.

The composable mirrors those backing variables back into reactive state and produces `resolvedPointsOfInterest`. When a live expression has no value yet (e.g. the source hasn't published), the POI falls back to its stored `fallbackCoordinates` and is flagged as having no valid position (driving the dimmed marker + tooltip).

Static POIs resolve their render position straight from their persisted definition rather than reading it back through the data lake, so edits like dragging apply immediately without a round-trip lag.

ArturoManzoli reviewed (changes requested):
Feature wise its working very well! 👍 

Some UI changes and we'll be good to go:

(1) Use the color `white` instead of `primary` for the switch;
(2) When switching  to dynamic, the coordinate values are left on the input field. They should be removed so it will be more intuitive for the users to see the next step on configuring the feature.
*  If you set your cursor to the beginning of the coordinate to select a data-lake variable, you are able to keep piling up string is the input field. Didn't test the effect on the data that the poi will receive, but interface wise this should be prevented.
* When switching back to non-dynamic POI, the data lake variables names are left on the input as well.

One more thing is to improve the text on the (?) helping icon for something more user friendly. Maybe a more descriptive text about how to use, and suggesting compound variables. The NaN expression might not be understood by many users.

<img width="556" height="777" alt="Image" src="https://github.com/user-attachments/assets/e1b01c6c-8187-4d1c-9ba1-d57801d0bc40" />

<img width="518" height="183" alt="Image" src="https://github.com/user-attachments/assets/4437e4df-367b-4ca1-bf6a-fa5adb7d8fbd" />

<img width="471" height="128" alt="Image" src="https://github.com/user-attachments/assets/2cb2b82b-5adb-4958-8f74-5283a62fd78c" />

<img width="525" height="704" alt="Image" src="https://github.com/user-attachments/assets/31e26335-f226-499d-82e5-aef68e9e7ef5" />

<img width="647" height="723" alt="Image" src="https://github.com/user-attachments/assets/2131f115-0cf0-4a8a-92ca-89fa982f0c18" />

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
## Review follow-up — round 1

### Done
- **`PoiManager.vue`** (@ArturoManzoli 1 — switch color): the "Dynamic positioning" switch now uses `color="white"` instead of `primary`.
- **`PoiManager.vue`** (@ArturoManzoli 2 — values linger when toggling mode, incl. variable names left when switching back to static): added a watcher that clears the latitude/longitude fields whenever the user toggles between static and dynamic (both directions). It's skipped during dialog setup so editing an existing POI still loads its values.
- **`PoiManager.vue`** (@ArturoManzoli help icon / "NaN"): the tooltip literally rendered `NaN` because `{{ mavlink / buoy / latitude }}` in the template was parsed by Vue as an interpolation (`mavlink / buoy / latitude` → division → `NaN`). Escaped the braces and rewrote the help text to explain usage and that variables can be combined with math, e.g. `{{ mavlink/buoy/latitude }} + 0.0001`.
- **`PoiManager.vue`** (Claude 1.1 / 6.1 — filler JSDoc): removed the `ccccc…`/`iiiii…` placeholder JSDoc; the `availableVariables` ref is now `ref<{ id: string; name: string }[]>([])`, which is self-describing.
- **`PoiManager.vue`** (Claude 1.2 — `showDialog` paired with `console.error`): removed the redundant `console.error` in the "POI not found" path and the matching one in `deletePoi`, since the dialog already surfaces the message.
- **`CompassHUD.vue`** (Claude 1.3 — home pseudo-POI with empty variable ids): added a comment explaining that Home is a static pseudo-POI with no data-lake backing, so the empty `latitudeVariableId`/`longitudeVariableId` are intentional and never read.
- **`ToolsMapView.vue`** (Claude 6.3 — deprecated deep selector): `::v-deep` → `:deep()`.
- **`utils-poi.ts`** (Claude 6.4 — unescaped tooltip HTML): added an `escapeHtml` helper and applied it to `poi.name`/`poi.description` in `getPoiTooltipHtml`.
- **`Map.vue`** (build fix): fixed `speedDialOpen is not defined` left over from extracting the center-on speed dial — wired `v-model:open` on `MapCenterControl` and declared the ref.

### Done differently
- **`PoiManager.vue`** (@ArturoManzoli 2 — "you can keep piling up strings" with the cursor at the start): the coordinate input is intentionally a free-form expression editor (so compound expressions like `{{ a }} * 10` work — which the new help text suggests). Clearing the field on mode-toggle removes the leftover value that made prepending feel like piling up. Happy to add hard input restrictions on top if you'd prefer.

### Deferred
- **`Map.vue` / `MissionPlanningView.vue`** (Claude 6.2 — duplicated POI marker management): agreed the marker `add/update/remove/sync` functions are near-identical and would make a good `usePoiMapMarkers` composable, but it's a pre-existing pattern (not a regression in this PR), so deferring to a dedicated follow-up to keep this PR scoped.

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
@ArturoManzoli all done.

ArturoManzoli reviewed (approved):
Everithing working!

rafaellehmkuhl commented:
/review

rafaellehmkuhl committed:
poi: add a Map tools menu with a points-of-interest manager

New Tools > Map submenu listing every POI with its status (static/live/no data)
and current coordinates, allowing users to add, edit and delete POIs via the
existing PoiManager dialog.

rafaellehmkuhl committed:
data-lake: cancel stale transforming-function initial evaluations on rebuild

Rebuilding the transforming-function listeners scheduled fresh 1s initial
evaluations without cancelling the previous ones. Those stale timeouts held an
outdated function reference and later wrote an old value back — e.g. reverting
one coordinate of a just-dragged POI, since updating a function replaces its
object and triggers a rebuild. Cancel pending initial evaluations before
rescheduling.

rafaellehmkuhl committed:
poi: render resolved live-tracked POIs on maps, HUD and edge arrows

Maps, the compass HUD, and POI edge arrows now consume resolvedPointsOfInterest
so live positions update automatically. Markers use static helper imports,
reflect status via color/opacity, and live-tracked POIs are not draggable.

rafaellehmkuhl committed:
poi: add pure helpers for marker color, opacity, tooltip and status

Live-tracked POIs without valid data are grayed out, dimmed, and show a
"Coordinates unknown" tooltip instead of their fallback position.

rafaellehmkuhl committed:
poi: edit coordinates with a data-lake variable picker

Each coordinate field accepts a fixed number or a data-lake expression, with a picker to insert any number-typed data-lake variable.

rafaellehmkuhl committed:
poi: model coordinates as static values or data-lake expressions

Redefine PointOfInterest so each coordinate is a PoiCoordinateSource
(number = static, string = data-lake expression) plus fallbackCoordinates,
and add ResolvedPointOfInterest as the runtime shape consumed by the UI.

rafaellehmkuhl committed:
poi: back coordinates with data-lake variables

Each POI exposes its position through cockpit/pois/<id>/latitude|longitude.
syncPoiCoordinateVariables keeps these in sync: static coordinates are written
directly, while expressions get a small resolver that listens to their
referenced variables and re-evaluates on change.

rafaellehmkuhl committed:
poi: manage points of interest via composable instead of mission store

Move POI ownership out of the mission store into a shared usePointsOfInterest
composable. It persists definitions, migrates legacy POIs, mirrors live
coordinate values from the data lake, and exposes resolvedPointsOfInterest
plus CRUD actions.

rafaellehmkuhl commented:
## Review follow-up — round 2

### Done
- **`MapCenterControl.vue` / `ToolsMapView.vue` / `PoiManager.vue`** (Claude 5.1 — new interactions lack `logUserAction`): added `logUserAction` in the owning handler for each new discrete interaction.
  - `MapCenterControl.vue`: center-on-mission, center/track home, center/track vehicle, center/track POI, and speed-dial open/close.
  - `ToolsMapView.vue`: center-on-POI, remove-POI.
  - `PoiManager.vue`: save (add vs. update), delete, toggle manual ID editing, and dialog open/close.
  - Dialog opens are logged once in `PoiManager.openDialog` (the single funnel the Tools list edit/new buttons and the map-marker click all reach) rather than in each caller, per the AGENTS.md "log at the funnel" guidance.

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
## Review follow-up — round 3

### Done
- **`CompassHUD.vue`** (1.1 — home pseudo-POI has empty `latitudeVariableId`/`longitudeVariableId`): extracted a dedicated `buildHomeHudPoi(coords)` helper that builds the synthetic home marker, and replaced the empty strings with clearly-invalid sentinels (`__home__/latitude`, `__home__/longitude`) so the required fields never read as real data-lake backing variables.

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
## Review follow-up — round 1

### Done
- **`src/components/mini-widgets/DepthIndicator.vue`, `RelativeAltitudeIndicator.vue`, `widgets/DepthHUD.vue`** (1.2 — `rawAltitude as number` cast without runtime guard): replaced the cast with a `typeof rawAltitude.value !== 'number'` early-return guard in all three widgets, so a non-numeric custom data-lake variable no longer silently produces `NaN`.
- **`src/composables/useAltitudeSourceConfig.ts` + the three widgets** (5.3 — missing `logUserAction`): added `onAltitudeSourceSelected` and `onUseCustomAltitudeVariableToggled` handlers in the composable (single funnel) and wired them via `@update:model-value` on the source select/autocomplete and the custom-variable checkbox.
- **`src/components/mini-widgets/DepthIndicator.vue`, `RelativeAltitudeIndicator.vue`** (2.2 — inconsistent `miniWidget` prop JSDoc): changed `"Mini widget reference"` to `"Configuration of the widget"` to match the established pattern (7 other mini-widgets).
- **`src/libs/data-sources/altitude.ts`** (1.1 — `rawAltitudeToMeters` silent pass-through): documented in the JSDoc that custom/non-preset variables are assumed to already be in meters and pass through unchanged.

### Won't change (with reasoning)
- **1.3 — `useResolvedDataLakeTemplate` stale `tick` on rapid template changes**: acknowledged as benign in the review; no user-visible effect, so leaving as-is.
- **4.1 — listener subscribed for full widget lifetime**: the callback just updates a ref that a computed filters; not worth the added open/close subscription churn at this scale.
- **5.2 — close button uses `<i class="mdi mdi-close">` vs `<v-icon>`**: this pattern already exists elsewhere (e.g. `BatteryIndicator.vue`, `VeryGenericIndicator.vue`), so it's consistent, not a divergence.
- **6.3 — `hudColor` hardcoded `'#FFFFFF'`**: keeping the literal; it's more robust than index-based `colorSwatches.value[2][0]` (and 9.1 flags index-based access as fragile).
- **6.4 — `String.fromCharCode(0x20)`**: pre-existing and consistent across both indicators; out of scope for this change.
- **9.1 — index-based default IDs (`altitudeSourceOptions[0]`/`[3]`)**: nit; the named exports (`defaultDepthAltitudeVariableId`, `defaultAltitudeIndicatorVariableId`) already give call sites a stable name.

### Deferred
- **Baro/pressure-altitude sources** (ES-Alexander): tracked in #2736.
- **MAVLink field descriptions in the source UI** (ES-Alexander): agreed as future work, best done after the data-lake variable-description infrastructure lands.
- **5.1 — config-dialog template duplication across the three widgets**: the logic is already shared via the composable; extracting a shared `AltitudeSourceConfigPanel.vue` is a larger refactor I'd rather do separately.

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
/review

rafaellehmkuhl commented:
## Review follow-up — round 4

### Done
- **`src/views/ToolsMapView.vue`** (1.3 — `removePoi` deletes with one click, no confirmation): the Tools-menu remove button now opens a `useInteractionDialog` warning/confirm before deleting the POI and its backing data-lake variables, consistent with other destructive actions.
- **`src/components/poi/PoiManager.vue`** (5.1 — longitude field lacked the help tooltip): mirrored the latitude field's help `ⓘ` tooltip onto the longitude fieldset, so both make clear they accept fixed values or data-lake expressions.

### Won't change (with reasoning)
- **7.1 — GNSS feature bundled into the POI commits**: false positive. The GNSS serial-source feature is PR #2815, which has merged into `master`; this branch is rebased on top of it. `git log origin/master..HEAD` shows only the 11 POI/data-lake commits — no `gnss:`/`sources:`/`console:` commits. The bot compared across the master-advance (`79ef9af1…9ce5359f`) and attributed master's GNSS to this PR (it noted the incremental diff includes other merged PRs).
- **1.1 (`ConsoleViewer` `activeLevels`), 1.2 (`gnss.ts` `ensureLinkListener`), 6.2 (`useGnss` disconnect log), 10.2 (`NmeaAggregator` tests)**: all belong to the GNSS subsystem (PR #2815, now in `master`), not to this PR's diff — out of scope here.
- **10.1 — redundant `defineModel` import in `MapCenterControl.vue`**: can't drop it. This repo's `.eslintrc.cjs` doesn't register `defineModel` as a known macro, so removing the import fails `no-undef`. Keeping the import.
- **4.1 — `syncPoiCoordinateVariables` O(n·m)**: acceptable at realistic POI counts (flagged as a nit / future-scaling note); not optimizing now.
- **6.1 — `openDialog` branch duplication**: the three branches differ in coordinate handling (edit vs. new-with-coords vs. new-without-coords); keeping them explicit reads more clearly than a merged branch. Nit, leaving as-is.

rafaellehmkuhl commented:
/review
